File: sdk-python/aegisagent/receipts.py
"""Verifiable action receipts (open format, scheme ``aegis-jcs-1``).

A receipt is a tamper-evident record of one agent-action decision. Receipts
form a per-tenant hash chain:

    receipt_hash = SHA-256(canonicalize(body))

where ``body`` is every receipt field EXCEPT ``receipt_hash`` and INCLUDES
``prev_receipt_hash``. Because the previous link is inside the hashed body,
altering any field or re-linking the chain is detectable. This module is the
reference verifier; the gateway emits receipts in the same format so auditors
can verify them independently of the runtime. See ``docs/action-receipt-spec.md``.
"""

from typing import Any, Dict, List

from .canon import canonical_hash

#: ``prev_receipt_hash`` value for the first receipt in a chain.
GENESIS_PREV = ""

RECEIPT_HASH_FIELD = "receipt_hash"
PREV_HASH_FIELD = "prev_receipt_hash"


def _body(receipt: Dict[str, Any]) -> Dict[str, Any]:
    """The hashed portion of a receipt: all fields except ``receipt_hash``."""
    return {k: v for k, v in receipt.items() if k != RECEIPT_HASH_FIELD}


def compute_receipt_hash(receipt: Dict[str, Any]) -> str:
    """SHA-256 of the canonical body (all fields except ``receipt_hash``)."""
    return canonical_hash(_body(receipt))


def seal_receipt(
    receipt: Dict[str, Any], prev_receipt_hash: str = GENESIS_PREV
) -> Dict[str, Any]:
    """Return a NEW receipt with ``prev_receipt_hash`` set and ``receipt_hash``
    computed. Does not mutate the input."""
    body = _body(receipt)
    body[PREV_HASH_FIELD] = prev_receipt_hash
    sealed = dict(body)
    sealed[RECEIPT_HASH_FIELD] = canonical_hash(body)
    return sealed


def seal_chain(
    receipts: List[Dict[str, Any]], genesis_prev: str = GENESIS_PREV
) -> List[Dict[str, Any]]:
    """Seal a list of receipt bodies into a linked chain. Returns new dicts."""
    chain: List[Dict[str, Any]] = []
    prev = genesis_prev
    for receipt in receipts:
        sealed = seal_receipt(receipt, prev)
        chain.append(sealed)
        prev = sealed[RECEIPT_HASH_FIELD]
    return chain


def verify_receipt(receipt: Dict[str, Any]) -> bool:
    """True if the receipt's stored ``receipt_hash`` matches its recomputed hash."""
    stored = receipt.get(RECEIPT_HASH_FIELD)
    if not stored:
        return False
    return compute_receipt_hash(receipt) == stored


def verify_chain(
    receipts: List[Dict[str, Any]], genesis_prev: str = GENESIS_PREV
) -> bool:
    """True if every receipt verifies AND each link references the prior
    receipt's hash (the first must reference ``genesis_prev``)."""
    prev = genesis_prev
    for receipt in receipts:
        if not verify_receipt(receipt):
            return False
        if receipt.get(PREV_HASH_FIELD) != prev:
            return False
        prev = receipt[RECEIPT_HASH_FIELD]
    return True

File: sdk-go/aegis/receipts.go
// Package aegis — verifiable action receipts (scheme aegis-jcs-1).
//
// A receipt is a tamper-evident record of one agent-action decision. Receipts
// form a per-tenant hash chain:
//
//	receipt_hash = SHA-256(canonicalize(body))
//
// where body is every field of the receipt EXCEPT receipt_hash, and INCLUDES
// prev_receipt_hash. Because the previous link is inside the hashed body,
// altering any field or re-ordering the chain is detectable.
//
// This file is the Go reference verifier; see docs/action-receipt-spec.md for
// the open receipt format and sdk-python/aegisagent/receipts.py for the Python
// reference implementation.
package aegis

import "github.com/lavkushry/aegisagent/sdk-go/canon"

const (
	// GenesisPrev is the prev_receipt_hash value expected for the first receipt
	// in a chain.
	GenesisPrev = ""

	receiptHashField = "receipt_hash"
	prevHashField    = "prev_receipt_hash"
)

// receiptBody returns the hashed portion of a receipt: all fields except
// receipt_hash.
func receiptBody(receipt map[string]any) map[string]any {
	body := make(map[string]any, len(receipt))
	for k, v := range receipt {
		if k != receiptHashField {
			body[k] = v
		}
	}
	return body
}

// ComputeReceiptHash returns the SHA-256 hex of the canonical body of receipt
// (all fields except receipt_hash). Returns an error if canonicalization fails.
func ComputeReceiptHash(receipt map[string]any) (string, error) {
	return canon.CanonicalHash(receiptBody(receipt))
}

// VerifyReceipt reports whether the receipt's stored receipt_hash matches its
// recomputed hash. Returns (false, nil) when the receipt_hash field is absent
// or empty; returns (false, err) when canonicalization fails.
func VerifyReceipt(receipt map[string]any) (bool, error) {
	stored, _ := receipt[receiptHashField].(string)
	if stored == "" {
		return false, nil
	}
	got, err := ComputeReceiptHash(receipt)
	if err != nil {
		return false, err
	}
	return got == stored, nil
}

// VerifyChain verifies a slice of receipts that form a hash chain. It returns
// true only when every receipt's receipt_hash verifies AND each receipt's
// prev_receipt_hash equals the prior receipt's receipt_hash (with the first
// receipt's prev_receipt_hash compared against genesisPrev, which is usually
// GenesisPrev/""). An empty chain returns true.
func VerifyChain(receipts []map[string]any, genesisPrev string) (bool, error) {
	prev := genesisPrev
	for _, receipt := range receipts {
		ok, err := VerifyReceipt(receipt)
		if err != nil {
			return false, err
		}
		if !ok {
			return false, nil
		}
		p, _ := receipt[prevHashField].(string)
		if p != prev {
			return false, nil
		}
		prev, _ = receipt[receiptHashField].(string)
	}
	return true, nil
}
